Add class and type for financial transactions #367

paroga opened the pull request:
This change introduces two new data types to group the financial
transactions. Now every transaction has a "type", which itself belongs
to a "class".
Types should be used add structured information to an transaction, instead
of writing it into the notice textfield. E.g. this could be used to have
different types depending on the source of money (cash vs. bank transfer).
Classes are shown as different columns in the tables and will be uses to
group transactions of specific types. They should be used if not the whole
amount of ordergroup should be used to order fodd. E.g. if there is a
deposit or membership fee, which is independent of the normal credit.
This will allow us to implement additional features based on classes in
the future. E.g. the sum of transactions in the "membership fee" class
must be positive to allow food orders or show a big warning if it is bellow
a certain value.

wvengen commented:
Wow, this is quite a change. I'm glad you explained it before :) I've skimmed through the code, will need to let it sink in a bit again - hope that's ok :)

wvengen commented:
Would you mind rebasing it? It's not even possible to run the code now (because of `Messagegroup`'s `type` attribute).

paroga commented:
I never thought that it will get accepted easily. :) But I want at least know if the direction is ok. If yes, I'll make a clean PR with translations and so on.

wvengen commented on app/controllers/finance/ordergroups_controller.rb:
```diff
@@ -1,20 +1,27 @@
+    select = ['groups.*']
+    FinancialTransactionClass.all.each do |c|
+      select << "sum(CASE financial_transaction_types.financial_transaction_class_id WHEN #{c.id} THEN financial_transactions.amount ELSE 0 END) AS sum_of_class_#{c.id}"
+    end
```
I wonder if this could become one SQL statement. Perhaps not if we want to keep compatible with MySQL and PostgreSQL. At the very least, we should move this to a helper method, or rather to the model.

wvengen commented on spec/integration/balancing_spec.rb:
```diff
@@ -1,6 +1,8 @@
 
 feature 'settling an order', js: true do
+  let(:ftc) { create :financial_transaction_class }
+  let(:ftt) { create :financial_transaction_type, financial_transaction_class: ftc }
```
creating a default transaction class can move to the transaction factory

paroga replied:
where is the transaction factory?

wvengen replied:
Ah, sorry - I meant something else. See comment in factory.

wvengen commented:
Thanks for rebasing! With regards to idea and UX, this came to mind:
* Rough idea looks fine.
* FTtypes would be useful for many.
* Let's have a set of useful defaults for FTtypes.
* Do we want a default FTclass or leave it empty?
* I wonder if we should access FTclasses in the configuration screen instead. Depends on how often we think people want to change it, how much can 'break' if people change it without fully understanding it.
* How will plugins be able to do stuff with specific FTtypes, e.g. online payment, imported bank transactions, etc. Also applies to the type for transactions created when an order is settled.
* Perhaps we need a full and abbreviated transaction class name (for table headers).

With regards to code, I noted:
* There's quite some stuff happening in the controllers. I think more could be moved to the model, or if that becomes to much, to concerns.
* Since it's financial transactions, I'm a bit more security-minded, and wonder if we should use strong parameters more.
* Transaction CSV download needs adaptation too (just to remember).

Good work! I'm curious how it will shape up.

paroga commented:
> Rough idea looks fine.

THX

> FTtypes would be useful for many.

not sure, if you do not have anything built on top of it, note is IMHO enough

> Let's have a set of useful defaults for FTtypes.

a) I do not see useful defaults (might be very specific per foodcoop)
b) If we add them the UI gets more complex (now I hide the type select everywhere when possible)
c) We would need translations for it, which is a problem for texts in the DB

> Do we want a default FTclass or leave it empty?

I think we should have one class and one type, so we do not need to handle NULL values

> I wonder if we should access FTclasses in the configuration screen instead. Depends on how often we think people want to change it, how much can 'break' if people change it without fully understanding it.

sounds reasonable. any good UI idea? having a ADD/DELETE/EDIT function for all of thats stuff might get complex very easy...

> How will plugins be able to do stuff with specific FTtypes, e.g. online payment, imported bank transactions, etc. Also applies to the type for transactions created when an order is settled.

I think that plugins need a config option, where admins can select FTtypes which should be used for specific actions.

> Perhaps we need a full and abbreviated transaction class name (for table headers).

I had it in the beginning, but removed it again, since I use name only (@ col header)

> There's quite some stuff happening in the controllers. I think more could be moved to the model, or if that becomes to much, to concerns.

should be better now

> Since it's financial transactions, I'm a bit more security-minded, and wonder if we should use strong parameters more.

I do not see a big problem since it's guarded with role_finance anyway.

wvengen commented on app/models/ordergroup.rb:
```diff
@@ -26,6 +26,17 @@ def non_members
+    end
+
+    select(columns.join(', '))
+      .joins('LEFT JOIN financial_transactions ON groups.id = financial_transactions.ordergroup_id LEFT JOIN financial_transaction_types ON financial_transaction_types.id = financial_transactions.financial_transaction_type_id')
```
split into two joins?

wvengen commented on app/views/finance/financial_transactions/_transactions.html.haml:
```diff
@@ -16,8 +16,15 @@
+      - FinancialTransactionClass.table_columns.each do |c|
+        %th
+          - if c.name?
+            = sort_link_helper c.name, "amount"
```
We could have `.name?` in the model fallback to the translation, would work on more places.

wvengen commented on app/views/finance/financial_transactions/new_collection.html.haml:
```diff
@@ -34,6 +34,12 @@
+      %b= heading_helper FinancialTransaction, :financial_transaction_type
+      = select_tag :type, options_for_select(FinancialTransactionType.order(:name).map { |t| [ t.name, t.id ] })
+  - else
+    = hidden_field_tag :type, FinancialTransactionType.default
```
:) `FinancialTransactionType.any?` could allow tiny database optimizations.

wvengen commented:
> > FTtypes would be useful for many.
>
> not sure, if you do not have anything built on top of it, note is IMHO enough

There could be a filter on FTtypes; no typos would be made, and people can just use one of the available types, requiring less instruction on what notes are used.
Building on top of it could also be default for settling an order.

> > Let's have a set of useful defaults for FTtypes.
>
> a) I do not see useful defaults (might be very specific per foodcoop)

True. Still a default set that applies to many could be helpful for starting foodcoops. Like payment and sale. And system transaction types like those created when settling an order (and, in plugins or in the future, automatic membership fee and online payments). Perhaps we could indeed start with one default transaction type, but still have system transaction types. But I'm running ahead of myself: let's just introduce transaction types first, then introduce the notion of system transaction types.

> b) If we add them the UI gets more complex (now I hide the type select everywhere when possible)

Agree.

> c) We would need translations for it, which is a problem for texts in the DB

Yep! Either in the database seed that would become language-dependent, or we add a code to transaction types. This could be used to identify system transaction types, as well as translations (which could perhaps be overridden by the name which would be empty by default).

> >  I wonder if we should access FTclasses in the configuration screen instead. Depends on how often we think people want to change it, how much can 'break' if people change it without fully understanding it.
>
> sounds reasonable. any good UI idea? having a ADD/DELETE/EDIT function for all of thats stuff might get complex very easy...

Not fully yet. I'm seeing a single index for both transaction classes and types, something like this (with some more explanation or so):

    Registration                  [edit] [delete]
    
        Registration Payment      [edit] [delete]
        Registration Refund       [edit] [delete]

    Membership                    [edit] [delete]
    
        Membership Payment        [edit] [delete]
        Membership Monthly Fee    [edit] [delete]

Than something like a "(Financial) Transaction Types" menu item would suffice.

> > How will plugins be able to do stuff with specific FTtypes, e.g. online payment, imported bank transactions, etc. Also applies to the type for transactions created when an order is settled.
>
> I think that plugins need a config option, where admins can select FTtypes which should be used for specific actions.

Hmm. I also would like it to work without having to configure this. Not sure if that's feasible.

> > Perhaps we need a full and abbreviated transaction class name (for table headers).
>
> I had it in the beginning, but removed it again, since I use name only (@ col header)
Ok. (Then we'd need to make the editbox for it small as well.)

> > Do we want a default FTclass or leave it empty?
>
> I think we should have one class and one type, so we do not need to handle NULL values

Agreed. In the db seed and in a migration. And not deletable.

> should be better now (regarding controller complexity)

Yes, thanks!

> I do not see a (security) big problem since it's guarded with role_finance anyway.

True.

paroga commented:
Can you check the new UI in my last commit? If it's ok, I'll fix the tests and add the missing checks to ensure that the last transaction type and class can not be removed.

paroga commented:
ping?

paroga commented:
can you look at this too?

wvengen commented:
Thanks for revisiting this! I need more time for review, hope to get back in a short while.

paroga commented:
np, just wanted to make sure, that you know that there is something new :-)

paroga commented:
Did you have already some time to look at this change? For me it would be very important go get at least a ok on the model changes. In that case it's easy to deploy this feature, since I do not have to worry about later database migrations...
If you want I can split the commit into smaller pieces to easy up the review process.

paroga commented:
I've split the commit into smaller parts to make the review easier.

wvengen commented on app/models/financial_transaction_class.rb:
```diff
@@ -0,0 +1,16 @@
+  validates :name, presence: true
+  validates_uniqueness_of :name
+
+  scope :table_columns, -> { order(name: :asc) }
```
I'd like to avoid default scopes, perhaps have a default `sorted` scope if you want to have a default sorting order. (Not `ordered` because we use it in a different sense elsewhere.)

wvengen commented on app/models/financial_transaction_type.rb:
```diff
@@ -0,0 +1,21 @@
+  validates_uniqueness_of :name
+  validates :financial_transaction_class, presence: true
+
+  before_destroy :check_last_financial_transaction_type
```
Maybe something like `restrict_deleting_last_financial_transaction_type` would be just a bit clearer?

wvengen commented on app/models/order.rb:
```diff
@@ -229,7 +229,7 @@ def finish!(user)
 
   # Sets order.status to 'close' and updates all Ordergroup.account_balances
-  def close!(user)
+  def close!(user, transaction_type)
```
This method is used by some (unofficial?) plugins, I'd recommend to revert to the default transaction type when not given.

paroga replied:
i've added support for a default type

wvengen commented on app/models/ordergroup.rb:
```diff
@@ -54,9 +54,9 @@ def get_available_funds(exclude = nil)
-  def add_financial_transaction!(amount, note, user)
-    transaction do      
-      t = FinancialTransaction.new(:ordergroup => self, :amount => amount, :note => note, :user => user)
+  def add_financial_transaction!(amount, note, user, transaction_type)
```
same here regarding a default transaction type fallback

wvengen commented on app/views/finance/financial_transactions/new.html.haml:
```diff
@@ -5,6 +5,10 @@
+  - if FinancialTransactionType.count > 1
+    = f.association :financial_transaction_type, :as => :radio_buttons
+  - else
+    = f.association :financial_transaction_type, :as => :hidden
```
Interesting that this works as a hidden field, didn't know.

wvengen commented on config/locales/de.yml:
```diff
@@ -1279,6 +1279,8 @@ de:
     delivery:
       each_stock_article_must_be_unique: Lieferung darf jeden Lagerartikel höchstens einmal auflisten.
+    financial_transaction_type:
+      no_delete_last: Es muss mindestens ein Finanaztransaktionstyp existieren.
```
typo?

wvengen commented on config/locales/en.yml:
```diff
@@ -1288,6 +1288,8 @@ en:
     delivery:
       each_stock_article_must_be_unique: Each stock article must not be listed more than once.
+    financial_transaction_type:
+      no_delete_last: At least one financital transaction type must exist.
```
typo?

wvengen commented on config/routes.rb:
```diff
@@ -148,7 +148,7 @@
 
           get :confirm
-          patch :close
+          post :close
```
ok

wvengen commented on db/seeds/minimal.seeds.rb:
```diff
@@ -21,5 +21,9 @@
 
+# First entry for financial transaction types
+financial_transaction_class = FinancialTransactionClass.create(:name => "Other")
+FinancialTransactionType.create(:name => "Foodcoop", :financial_transaction_class_id => financial_transaction_class.id)
```
Is it required to have a financial transaction type? Else I'd leave it out in the seeds.
It would be nice to have a database seed showcasing more extensive use of financial transaction classes.

paroga replied:
financial transaction type is mandatory

wvengen replied:
Ah, I understand now.

wvengen commented on spec/factories/financial_transaction_type.rb:
```diff
@@ -0,0 +1,13 @@
+  end
+
+  factory :financial_transaction_type do
+    sequence(:name) { |n| Faker::Lorem.words(rand(2..4)).join(' ') + " ##{n}" }
```
Since a type isn't valid without a class, I'd add the line `financial_transaction_class` to it. See [FactoryGirl#associations](http://www.rubydoc.info/gems/factory_girl/file/GETTING_STARTED.md#Associations).

paroga commented:
i've updated the first commit, cleaned it a little bit up and rebase the other two commits
ad seeds) yes, i like to have also more stuff but (a) i don't want to make this feature too visible in the first step, (b) don't want to change existing behaviour (people do not need to care about the type ATM) and (c) what about translations? IMHO we can make this in a later step...

wvengen commented on app/models/ordergroup.rb:
```diff
@@ -54,9 +54,10 @@ def get_available_funds(exclude = nil)
+  def add_financial_transaction!(amount, note, user, transaction_type = nil)
+    transaction_type = FinancialTransactionType.default if transaction_type.nil?
+    transaction do
+      t = FinancialTransaction.new(:ordergroup => self, :amount => amount, :note => note, :user => user, :financial_transaction_type => transaction_type)
```
What about moving the default transaction type to `FinancialTransaction` itself? In that way plugins creating a financial transaction using the constructor (I've done it at some point, but not sure if that survived or not in the final plugin), would still work. Perhaps an `after_initialize` there to set the default transaction, documenting that it is for backward compatibility.
What do you think?

paroga replied:
i like the idea and updated the commit

wvengen commented:
I'll continue with the other commits later. Thanks for your patience!

paroga commented:
ping?

wvengen commented on app/views/finance/financial_transactions/_transactions.html.haml:
```diff
@@ -16,14 +16,23 @@
       - if with_ordergroup
         %th= heading_helper FinancialTransaction, :ordergroup
       %th= heading_helper FinancialTransaction, :user
+      - if FinancialTransactionType.count > 1
```
`.any?` (more simple database call)

paroga replied:
any is not >1 it's >0

wvengen replied:
oops I was sleeping :o, sorry

wvengen commented on app/models/financial_transaction_class.rb:
```diff
@@ -3,4 +3,14 @@ class FinancialTransactionClass < ActiveRecord::Base
+  scope :sorted, -> { order(name: :asc) }
+
+  def display
+    if FinancialTransactionClass.count > 1
```
`any?` can result in a simpler db-call

wvengen commented on app/views/home/index.html.haml:
```diff
@@ -53,15 +53,24 @@
         %tr
           %th= heading_helper FinancialTransaction, :created_on
           %th= heading_helper FinancialTransaction, :user
+          - if FinancialTransactionType.count > 1
```
same here

wvengen commented on app/views/home/index.html.haml:
```diff
@@ -53,15 +53,24 @@
           %tr
             %td= format_time(ft.created_on)
             %td= h(show_user(ft.user))
+            - if FinancialTransactionType.count > 1
```
same here

wvengen commented on app/views/home/ordergroup.html.haml:
```diff
@@ -9,6 +9,11 @@
         %b= heading_helper(Ordergroup, :available_funds) + ':'
         = number_to_currency(@ordergroup.get_available_funds())
+      - if FinancialTransactionClass.count > 1
+        - FinancialTransactionClass.sortd.each do |c|
```
typo?

wvengen commented on app/views/home/ordergroup.html.haml:
```diff
@@ -9,6 +9,11 @@
       %p
         %b= heading_helper(Ordergroup, :available_funds) + ':'
         = number_to_currency(@ordergroup.get_available_funds())
+      - if FinancialTransactionClass.count > 1
```
same here

wvengen replied:
What about introducing a method on `FinancialTransactionClass` that tells whether the financial transactions class feature is needed?

paroga replied:
+1, any good name?

wvengen commented on app/controllers/admin/financial_transaction_classes_controller.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def new
+    @financial_transaction_class = FinancialTransactionClass.new(params[:financial_transaction_class])
+    render :layout => false
```
I would like to transition to new-style options `layout: false`, any objections?

paroga replied:
agree

wvengen commented on app/controllers/admin/financial_transaction_classes_controller.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def create
+    @financial_transaction_class = FinancialTransactionClass.new(params[:financial_transaction_class])
+    if @financial_transaction_class.valid? && @financial_transaction_class.save
```
I think that [`save`](http://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-save) already checks for validity, and returns that.

wvengen commented on app/controllers/admin/financial_transaction_classes_controller.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def destroy
+    @financial_transaction_class = FinancialTransactionClass.find(params[:id])
+    @financial_transaction_class.destroy
```
`destroy!`?

wvengen commented on app/controllers/admin/financial_transaction_types_controller.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def create
+    @financial_transaction_type = FinancialTransactionType.new(params[:financial_transaction_type])
+    if @financial_transaction_type.valid? && @financial_transaction_type.save
```
same as before

wvengen commented on app/controllers/admin/financial_transaction_types_controller.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def destroy
+    @financial_transaction_type = FinancialTransactionType.find(params[:id])
+    @financial_transaction_type.destroy
```
same here `destroy!`?

paroga commented:
new commits pushed

paroga commented:
ping?

wvengen commented on app/controllers/admin/financial_transaction_classes_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  def create
+    @financial_transaction_class = FinancialTransactionClass.new(params[:financial_transaction_class])
+    if @financial_transaction_class.save
+      redirect_to update_transaction_types_admin_finances_url, status: 303
```
I don't think we use 303 status elsewhere, seems like a good idea to me.

wvengen commented on db/migrate/20160219123220_create_financial_transaction_class_and_types.rb:
```diff
@@ -0,0 +1,22 @@
+
+    execute "INSERT INTO financial_transaction_classes (id, name) VALUES (1, 'Other')"
+    execute "INSERT INTO financial_transaction_types (id, name, financial_transaction_class_id) VALUES (1, 'Foodsoft', 1)"
+    execute "UPDATE financial_transactions SET financial_transaction_type_id = 1"
```
this isn't really reversible, is it?

wvengen replied:
```
$ rake db:migrate:down VERSION=20160219123220
== 20160219123220 CreateFinancialTransactionClassAndTypes: reverting ==========
rake aborted!
StandardError: An error has occurred:

ActiveRecord::IrreversibleMigration
```

wvengen commented:
I think in terms of functionality, it's there. Great!
In terms of user-experience, I think there is some way to go. Right now, as a user seeing the new screen, I have no clue what to dot. Adding explaining texts would already help. It is a bit of a difficult concept, so I'm not really sure yet how to keep it short and clear at the same time. But let's see where we get.

Small remarks and ideas that could improve it:
* _Do you really want to delete ..._ could use an indication of what kind is being deleted, "_Do you really want to delete Foodsoft?_" can be a bit frightening.
* Does there need to be a default transaction type for settling an order?
* I don't think the user needs to fully understand _class_ and _type_, UX could help a lot here. E.g.
  - a `+` button behind the class name to add a new child type
  - above the table an 'Add new class' button. 
* What happens when a financial type is deleted? Should it be possible when there are transactions linked to it?
* Perhaps _Default_ is a more suitable name for the transaction instance created on migration?
* Do you think a description for each class (and maybe type) would add comprehensibility? Person setting up the classes would enter them. Could be shown in small font below the name.
* Could also add a user-visible description on hover in account statement to explain. (But if you don't think this is necessary, please ignore.)
* Just a reminder: don't forget to complete i18n :)
* Nice to have: provide a link to the financial transactions list of each type

All in all, good work. Thanks!

paroga commented:
> this isn't really reversible, is it?

It's reversible if you delete the new tables and the column.

> In terms of user-experience, I think there is some way to go. Right now, as a user seeing the new screen, I have no clue what to dot. Adding explaining texts would already help. It is a bit of a difficult concept, so I'm not really sure yet how to keep it short and clear at the same time. But let's see where we get.

It's the i18n for the "paragraph"...

> Do you really want to delete ... could use an indication of what kind is being deleted, "Do you really want to delete Foodsoft?" can be a bit frightening.

IMHO getting rid of all the different delete messages in the past was a good idea. I don't like to reintroduce them.

> Does there need to be a default transaction type for settling an order?

You need at least one transaction type, which will be hidden if it's only one.

> a + button behind the class name to add a new child type

Ok for me.

> What happens when a financial type is deleted? Should it be possible when there are transactions linked to it?

ATM you can not delete a type, which is already used.

> Perhaps Default is a more suitable name for the transaction instance created on migration?

"Default" is a language-specific term, but ok for me.

> Do you think a description for each class (and maybe type) would add comprehensibility?

IMHO an improvement for the next steps. (I didn't want to make the already big change even bigger^^)

> Just a reminder: don't forget to complete i18n :)

No, as written in the first statement. It's also necessary for explanation.

> Nice to have: provide a link to the financial transactions list of each type

An additional improvement for a coming PR.

wvengen commented:
Thanks for the answers. Sounds good. Remaining:

> > Does there need to be a default transaction type for settling an order?
> 
> You need at least one transaction type, which will be hidden if it's only one.

If there are multiple transactions types, it might be useful to have a default (not just the first). But this can be an improvement later, if needed.


> "Default" is a language-specific term, but ok for me.

If this only applies to English, that's ok. "Other" sounds like "things outside of the ordinary" to me, while "Default" is more like "unless you have a better idea, this is what to use".



> >  Do you think a description for each class (and maybe type) would add comprehensibility?
> 
> IMHO an improvement for the next steps. (I didn't want to make the already big change even bigger^^)

:+1:

paroga commented:
>> "Default" is a language-specific term, but ok for me.

>If this only applies to English, that's ok. "Other" sounds like "things outside of the ordinary" to me, while "Default" is more like "unless you have a better idea, this is what to use".

I mean that it would require different database migrations for each language. (e.g. is should be "Standard" in German)

wvengen commented:
Hmmm ... that's a good one. Wouldn't _Standard_ make sense, as it is similar to some other languages?
For database migrations, I think getting the language (not using the settings infrastructure because that may change in the future, but by a direct query) and choosing one of the translations defined in the migration (because also locales may change) would actually work.
Now database seeds may be localized as well (though they really aren't right now, and one cannot provide a language for database creation). So for the future, translations would be ok, I think.

paroga commented:
I've updated the PR. Fixed the migration, changed the default class name and adopted the UI.
IMHO the only missing part ist the I18N now (which I don't want to do more than once). Can you crosscheck that? If you agree with that I'll add the translations and hope to finish this PR then finally. :-)

wvengen commented on app/models/financial_transaction.rb:
```diff
@@ -3,15 +3,20 @@
   localize_input_of :amount
 
+  after_initialize do
+    self.financial_transaction_type = FinancialTransactionType.default unless self.financial_transaction_type
```
Shorter would be
```ruby
self.financial_transaction_type ||= FinancialTransactionType.default
```

wvengen replied:
Moving this to a (private) method and using `after_initialize` with a symbol would make it just even a bit more flexible (allow a plugin to override it).

wvengen commented on app/models/financial_transaction_type.rb:
```diff
@@ -0,0 +1,25 @@
+  validates_uniqueness_of :name
+  validates :financial_transaction_class, presence: true
+
+  before_destroy :restrict_deleting_last_financial_transaction_type
```
:+1:

paroga commented:
@wvengen ping?

wvengen commented:
Would prefer to look at this after the foodcoops.net transition is done, ok?

paroga commented:
sure, just wanted to ping, for the case you didn't see the new commits...

wvengen commented:
Thanks, I hadn't seen them yet :)

paroga commented:
4.6 released, so it should be ok to merge it into master finally?

wvengen commented:
What's going on with the specs?

wvengen commented:
What's going on with Travis?
Also, did you consider where to write specs for? It adds complexity, and would be good for new contributions to not break this inadvertently.

paroga commented:
It mainly adds complexity to the views and not to the model, so I do not see a lot of code to test. For what do you want a test for?

I'll check travis, if the rest is fine.

wvengen commented:
After going through the code again, I agree that there's not actually that much new functionality.
I don't have this PR fully clear in my head, but I'll try: maybe `include_transaction_class_sum` or is that really internal stuff?

wvengen commented:
Other than these, I'd say, go!

paroga commented:
i've added a test for `include_transaction_class_sum` and fixed travis. imho it should be ready to merge now

paroga commented:
@wvengen ping

paroga commented:
@wvengen ping
